### DISCUSSION TITLE
Vite as Quasar build tool

### DISCUSSION THREAD

**Discussion by stefanvanherwijnen** [quasarframework/quasar#11587] [Ideas / Proposals]
# Vite support is now available in [@quasar/cli](https://quasar.dev/start/quasar-cli).

My project is available at https://github.com/simsustech/vitrify
<details>
<summary>Original post (outdated)</summary>
Hi,

For the past few months I have been working on [quasar-app-vite](https://github.com/stefanvanherwijnen/quasar-app-vite/), a Vite based alternative for @quasar/app. In contrast to the [Vite plugin](https://quasar.dev/start/vite-plugin), this is a standalone build tool for Quasar projects with support for SSR and Quasar specific features (app extensions, boot files, etc).

With the basics now functioning, I am looking for some feedback.

## Background

While developing app extensions, I ran into problems with build and bundle tools. The app-extension-ui starter kit provided some help, but I got stuck on trying to add TypeScript support. Eventually I switched to using Vite, and everything magically worked out of the box. In the mean time, ES modules support in Node matured and the combination of ESM, Vite and pnpm already saved me a lot of time debugging and fixing annoying errors like `instance is null` (multiple Vue instances in the project due to for example an app extension having `vue` as devDependency).

## High level overview

Everything is written in TypeScript with ES module syntax, and the package manager used is pnpm. pnpm is a strict package manager, i.e. every package used is required to be a direct dependency of the project (i.e. the way quasar uses @quasar/app for the vue dependency will not work for pnpm). This will be confusing at first because dependencies of dependencies will not be resolved in your project, but it is much more transparent which dependency will be used when you import one: the version defined in `dependencies` in package.json.

quasar-app-vite uses the index.html in your project as entry point. The JS entry point is injected at runtime by [QuasarPlugin](https://github.com/stefanvanherwijnen/quasar-app-vite/blob/master/packages/quasar-app-vite/src/vite-plugin-quasar.ts#L49):

- ssr/entry-client.ts for SSR client builds
- ssr/entry-server.ts for SSR server builds
- csr/entry.ts for client-side rendered builds (e.g. standard SPA's or PWA's)

The entry script performs the correct setup for the build mode. The Vue app is initialized with main.ts and checks if the build is a SSR or CSR build, and Quasar is initialized.

quasar.conf.js can be used to setup the Quasar Vue plugin. `framework.components`, `framework.plugins`, `boot` and `extras` are parsed and imported in main.ts as a [virtual file](https://vitejs.dev/guide/api-plugin.html#importing-a-virtual-file) (instead of [statically replacing the content with templates](https://github.com/quasarframework/quasar/blob/9465d9c437084b01bd6ca36a35233c3ac0164694/app/templates/entry/quasar-user-options.js)).

You can use:

- `quasar-vite build`: Build your Quasar project
  - `-m`: Build mode: `csr` or `ssr`
  - `--base`: For SSR: https://vitejs.dev/guide/build.html#public-base-path
  - `--outDir`: Output directory, defaults to `./dist`
- `quasar-vite dev`: Spin up a development server

  - `-m`: Build mode: `csr` or `ssr`
  - `--host`: https://vitejs.dev/config/#server-host

  The SSR dev server will spin up a Fastify server with a Vite dev server in [middleware mode](https://vitejs.dev/config/#server-middlewaremode).

The QuasarPlugins roughly consists of the following:

- a HTML transform plugin to inject the correct entry point for the build mode
- a unplugin-vue-components plugin for auto-importing Quasar components in SFC's. The paths to Quasar components are resolved within an [importmap](https://github.com/stefanvanherwijnen/quasar-app-vite/blob/master/packages/quasar-app-vite/src/import-map.ts)
- a plugin to merge the `vite` entry in quasar.conf.js
- a plugin to provide virtual modules:
  - 'virtual:quasar-conf':
    - quasar.conf.js
  - 'virtual:quasar-extensions':
    - quasar.extensions.json
  - 'virtual:quasar-plugins':
    - JS plugins from `framework.plugins`
  - 'virtual:quasar-components':
    - JS components from `framework.components`
  - 'virtual:quasar-extras':
    - @quasar/extras from `extras`
  - 'virtual:quasar-boot':
    - JS files from the boot folder
  - 'virtual:fastify-setup':
    - The setup function from `fastify.setup`
- Finally a plugin to inject all that is required for Quasar:
  - aliases for app and quasar paths.
  - [build options](https://vitejs.dev/config/#build-options) for SSR
  - `define` for defining the [Quasar platform variables](https://github.com/quasarframework/quasar/blob/dev/ui/src/plugins/Platform.js)
  - `css` for injecting the required CSS/SASS.

The main problems at this time are:

- SSR builds are required to be build from source, as Quasar does not have a SSR bundle. Ideally (but maybe not possible), one would be able to perform `import { ... } from 'quasar'` in a NodeJS file and it will serve a build with the correct SSR platform variables. For now, everything is defined as [external](https://vitejs.dev/guide/ssr.html#ssr-externals) which means that the resulting SSR bundle will not contain any imports other than Node built-ins, express and fastify.
- @quasar/app uses require() for the app extension API, which works for CJS bundles but not for ESM and quasar-app-vite uses import() which works for ESM but not for CJS. Also, all webpack related features of the API are depcrecated.
- `import 'quasar/src/css/index.sass'` is the easiest way to load all CSS, but I would prefer to have tree-shaken CSS and only import what is required. However, for example the [Dialog plugin](https://github.com/quasarframework/quasar/blob/dev/ui/src/plugins/Dialog.js) imports the DialogPlugin component which is not automatically resolved in the importmap. Thus we need to explicitely import `components/dialog-plugin/DialogPlugin.sass`. This could be solved if the source code uses [@import](https://vitejs.dev/guide/features.html#import-inlining-and-rebasing) such that CSS is automatically resolved.
- Capacitor/Cordova/Electron builds are not working yet but should be possible with Vite plugins.
- All webpack related features will have to be replaced if still required. This basically means a major redesign of the `app` API.

Some things to discuss:

- What changes to the UI source code will be required and allowed?
- How to support SSR bundles? Possibly by replacing static platform variables with env variables (import.meta.env, if undefined it is a browser which is importing the code).
- API design for app extensions and the CLI. I would prefer more modularity. E.g. it would be nice if the [render()](https://quasar.dev/app-extensions/development-guide/install-api#api-render) function would allow a `renderer` argument so that the developer has more freedom (and if not provided it will use a default renderer). Modularity for the CLI can be achieved with Vite plugins which can be added to `vite.plugins` in quasar.conf.js

So I am looking for some feedback from the team and other Quasar users. In my humble opinion I would drop all UMD and CJS support and use Vite as a bundler for Quasar in the future. This obviously is a huge change and any current projects will not be compatible. Before continuing it would be great if we could collectively decide how to redesign the API, which issues and PR's to include or fix and which general direction we should take with Vite.

Anyway, I will definitely continue my work with Vite and any feedback or guidance is more than welcome.

- Stefan
</details>

**Comment by IlCallo** [comment]
This is really impressive! Thanks for your work in this area :)
As for all major and breaking changes (eg. AE API changes, dropping CJS, etc), we'll need to go through the [RFC process](https://github.com/quasarframework/rfcs)

For the AE API changes, @hawkeye64 should put his weight in, as he's the maintainer of most of Quasar first-party AEs and also did some tests with Vite. Could be we make both "modes" and "AEs" Vite plugins? That could be an interesting path to check for 

I'll also ping someone from Vite community to jump in the discussion, I guess they can provide us some insights on how to solve the problems you found, and define which things will need to change into Quasar core to support all of this, possibly without a major breaking change for the whole Quasar ecosystem

**Reply by stefanvanherwijnen** [reply]
Moving all extra features ("modes" and "AEs") to Vite plugins would definitely simplify things.

One thing to consider is removing `quasar.conf.js` and `quasar.extensions.json` and make both of them arguments for the Vite plugin. But again, this will break all backward compatibility. However, if the new base API is concise and modular, any legacy support could be added with more Vite plugins.

Also maybe it is possible to outsource a lot of the AE functionality to the package manager. I.e. `hasPackage` performs a manual check in package.json, but this should be the responsibility of the package manager. `peerDependencies` kind of is an alternative, but it only warns the user about missing packages instead of throwing an error.

At this point trying to provide support for older projects and app extensions is the main issue. As in the OP, working with ES and CJS in the same node package is impossible, so a redesigned `InstallAPI` will definitely not work with older AE's. I think making AE's Vite Plugins is the way to go so I'll see if I can get it to work.

**Reply by stefanvanherwijnen** [reply]
@IlCallo Do you know if there is a reason why sass is not imported at component level (e.g. `import ./QBtn.sass` in QBtn.js)

**Reply by pdanpdan** [reply]
The main reason was Qv1 - it had both .sass and .styl files for each component, so the import was not practical.
And in Qv2 the import was not added :(

**Reply by stefanvanherwijnen** [reply]
Correct me if I'm wrong, but since postcss-purgecss is problematic with quasar, the only way to 'tree-shake' css is by specifically importing it. In order not to miss css of local imports (e.g DialogPlugin imports Dialog), we either need to import sass in QDialog or add a DialogPlugin.sass.

Edit: Maybe some experienced sass developer could have a look at this: https://www.smashingmagazine.com/2019/08/handling-unused-css-sass-performance/.

**Reply by mila76** [reply]
For css 'tree-shake' i like the unocss style. If quasar embrace this way (Or maybe only on a plugin way) the css part can reach a total new level

I used your first attempt on vite with quasar (I think I see your post on reddit) months ago and worked perfectly for my small mpa needs

After the official vue plugin come out I removed the unplugin-vue-components and now work in production too. 

Quasr with Vite Is a good improvement in many way but I miss small little detail from webpack. Someday maybe all fixes with your work and quasar teams

**Reply by smolinari** [reply]
@mila76 
> but I miss small little detail from webpack.

What details? The improvements can only be made, if your wishes are known. Also, I'd suggest you put them in Stefan's repo as issues (if you haven't already). 

Scott

**Reply by mila76** [reply]
I can't configure a dir for fonts different from css. Not a big deal. Maybe is already doable but info on Vite and roolup is less than webpack

**Reply by stefanvanherwijnen** [reply]
@mila76 Anything you've seen on Reddit wasn't mine :).
As Scott said, let us know what you're missing. That's the reason for this discussion after all.

I'll have a look at UnoCSS. I happened to come across a YT video of Luke yesterday about it, but never heard of it before.

**Reply by mila76** [reply]
Sorry @stefanvanherwijnen .... that guy used a similar idea of yours 👯 
In #6775 we talked about unoCss...
My idea is to use css for the utility part of Quasar... need to create a presets like this (just first classes converted)

`{
      presets: [],
      theme: {
        colors: {},
        size: {
          xs: '4px',
          sm: '8px',
          md: '16px',
          lg: '24px',
          xl: '48px',
        },
      },
      rules: [
        ['no-shadow', { 'box-shadow': 'none !important' }],
        ['q-ma-auto', { margin: 'auto' }],
        ['q-pa-auto', { padding: 'auto' }],
        [/^q-m[xy]-(\d+)$/, ([, d]) => ({ 'margin-left': `${d / 4}rem`, 'margin-right': `${d / 4}rem` })],
        [/^q-ma-(\d+)$/, ([, d]) => ({ margin: `${d / 4}rem` })],
        [/^q-p[xy]-(\d+)$/, ([, d]) => ({ 'padding-left': `${d / 4}rem`, 'padding-right': `${d / 4}rem` })],
        [/^q-pa-(\d+)$/, ([, d]) => ({ padding: `${d / 4}rem` })],
        [
          /^q-p[xy]-(.*)$/,
          ([, c], { theme }) => {
            if (theme.size[c]) return { 'padding-left': theme.size[c], 'padding-right': theme.size[c] };
          },
        ],
        [
          /^q-pa-(.*)$/,
          ([, c], { theme }) => {
            if (theme.size[c]) return { padding: theme.size[c] };
          },
        ],
      ],
    }`

Uno css create a build with only the utility class used in project and you can create tons of other without making bundle super eavy

**Comment by smolinari** [comment]
A question I have is would using a Vite powered CLI, as you have currently developed it, exclude other package managers i.e. Yarn and NPM @stefanvanherwijnen? 

TBH, I personally wouldn't care, as I'm using PNPM myself with Quasar currently and as you described, I had to pull all of q/app's dependencies into my project to make it work with PNPM. But, I also have a Rush monorepo including a backend app and am happy with it all working. I'm just thinking about other devs, who might not want to use PNPM. 

Edit: on a more personal note, it is really awesome for you to take this on. It is the next level for Quasar too. So 😎 !!

Scott

**Reply by IlCallo** [reply]
I think latest Yarn and NPM behave (or can behave) similar to pnpm, it's just the "old generation" of package managers which doesn't support pnp properly

**Reply by stefanvanherwijnen** [reply]
I haven't really tested it, but in theory the package manager should not matter. I believe Vite itself also supports yarn and npm, so it should work.

The main issue I encounter with yarn is when linking packages, the linked package uses the `vue` instance from its node_modules, instead of the main `vue` instance in the app's node_modules (resulting in `instance is null`). The only workaround I found was using yalc, but that is quite cumbersome.

At this point its mainly trying to put all the pieces together to try and get the basics working, and I try to avoid any package manager related issues. Definitely something to keep in mind for the future though.

**Reply by hawkeye64** [reply]
For `yarn`, I think you would use the `nohoist` command: https://classic.yarnpkg.com/blog/2018/02/15/nohoist/

**Reply by smolinari** [reply]
It's funny to hear I wasn't alone with running into these kinds of issues with other package managers. I thought it was just my silly inexperienced self. 😜 I'm very stoked now to work with the Quasar-Vite CLI. Did you say it is ok to use now? For like a simple project with only one AE being used? I'm in POC mode anyway, I could try it out @stefanvanherwijnen.

On the backwards compatibility question, I'm of the opinion that Vite offers enough advances and advantages, that breaking changes, even serious ones, are worth it. That is just me though. 😊 

Scott

**Reply by stefanvanherwijnen** [reply]
If you clone the repo you can give the [example project](https://github.com/stefanvanherwijnen/quasar-app-vite/tree/master/packages/example)  a try. Note that the QCalendar AE does not work out of the box, but is manually initialized with a boot file.

I'll refractor things in the coming days and just use the global css import for example. Most issues at the moment are from missing imports. When that is done it should definitely work for simple projects. Right now probably some plugins aren't working properly.

**Reply by smolinari** [reply]
@stefanvanherwijnen - The AE I'm using is the Apollo AE to connect to a backend. If I find the time, I'll take a look.

Scott

**Comment by Shinigami92** [comment]
Hello @stefanvanherwijnen.
I'm from the Vite Team (not directly a core >code< contributor but more like someone that triages issues and PRs and making reviews)
I will get a new job next month (january 2022) and there I will build up a Vue 3 + Quasar + Electron Desktop Application. So from then on I will switch from Vuetify with Vue 2 to the new world :)
I hope I can start help after I arrived into the new ecosystem and help Quasar CLI get stable with Vite. My main-goal would be to get rid of webpack and make it possible to use Vite + Quasar + Electron.
_Sadly_ I'm not really interested / into SSR at all. I'm more specialized on SPAs.

If you want to get in contact with me feel free to find me on Quasar and/or Vite Discord Server.
Hopefully I could help managing some contacts between the projects so we can bring the community further.

**Reply by hawkeye64** [reply]
> ...I will get a new job next month (january 2022) and there I will build up a Vue 3 + Quasar + Electron Desktop Application....

@Shinigami92 You may be interested in this project: https://github.com/hawkeye64/electron-quasar-file-explorer-v2

**Reply by Shinigami92** [reply]
I already get my hands dirty with this hobby project: https://github.com/Shinigami92/watchlist

**Reply by hawkeye64** [reply]
Nice. I could possibly use that. I'll review it later. Starred for now.

**Reply by stefanvanherwijnen** [reply]
One of the reasons I started this project is because I was confused about SSR. You can develop SPA's just fine with SSR, it's just the initial rendering is performed on the server instead of your browser. If you don't use SSR mode then everything works as normal (client-side rendered)
When Vite 2.7.0 is released SSR support should be stable, but with the latest beta everything seems to work already.

One thing I am personally curious about is if [tauri](https://github.com/tauri-apps/tauri) could replace Electron. I have never developed with Electron, but the programs that use it always seem memory-heavy :disappointed: . I think it's already possible to compile a Vite app with Tauri, but I haven´t tried it myself.

**Reply by smolinari** [reply]
I'd consider Tauri as a possible "next advancement". Definitely worth looking at, but something for later. 😃 We need the CLI running first. 

Scott

**Reply by hawkeye64** [reply]
@stefanvanherwijnen 
> One thing I am personally curious about is if [tauri](https://github.com/tauri-apps/tauri) could replace Electron. I have never developed with Electron, but the programs that use it always seem memory-heavy disappointed . I think it's already possible to compile a Vite app with Tauri, but I haven´t tried it myself.

I have used Tauri with Quasar. It was a 20-second integration.

**Reply by Shinigami92** [reply]
For now, Tauri is to new for me and I will only use it when it reached some more importance. Also I think many in depth stuff that already could work in Electron would not work in Tauri just due to it's to new and got rid of old stuff.
But let's not further talk here about Tauri, I think this doesn't have something to do with Vite + Quasar

**Reply by stefanvanherwijnen** [reply]
@Shinigami92 If you could get an electron vite plugin (e.g. https://www.npmjs.com/package/vitejs-plugin-electron)  working with a Vue project, integrating Quasar should be easy. A separate plugin for electron would be preferable, but most electron+vue projects seem to use a different approach, not sure why (never used electron).

**Comment by stefanvanherwijnen** [comment]
I refactored the code and simplified a lot of things and released 0.7.0.
QuasarPlugin now has two arguments: `quasarConf` and `quasarExtensionIndexScripts` for quasar.conf.js and the index scripts of AE's respectively.
The main problem now is how to support app extensions. Basically it can be split in two:
- The index scripts modify things at build time and can in theory be moved to a Vite plugin
- The other scripts require their own CLI functionality, which at the moment is not included.

Also, for SSR to function properly, all AE's need to be ES modules ("type": "module"). The reason for this is that Vite copies all code in CJS packages as-is. So any boot files inside the AE are not transpiled and will result in `missing import at 'import ... from 'quasar-ui-...'` when running the SSR bundle. This means that none of the current AE's based of the starter-kits will work with SSR+Vite.

I'm wondering if @hawkeye64 has any ideas on how to best support AE's in Vite? Also, maybe someone from the Vite community knows if it is possible to add a custom entry to the Vite config (namely QuasarConfig)? That would make it much easier to modify the AE index scripts to run as Vite plugins.

**Reply by hawkeye64** [reply]
@stefanvanherwijnen Regarding AEs, I won't have time until the new year to help out with this and I would love to get involved.

**Comment by armenr** [comment]
@stefanvanherwijnen - Fantastic work! This is incredible.

A quick question about capacitor/cap modes...am I correct in understanding that it isn't supported yet?

Example:
- Project bootstrapped with vite
- Install & configure Quasar/quasar plugin
- Successfully render a Quasar Notification button inside the vite app (successfully) to verify that it's working

^^ In a project like this, you cannot, yet do quasar mode add capacitor, correct?

**Reply by hawkeye64** [reply]
That's correct. Until we have the option of Vite or Webpack in a quasar project, you cannot do the other modes without Quasar CLI.

**Reply by stefanvanherwijnen** [reply]
A simple notification should work fine. Capacitor mode should somehow be added as a Vite plugin, but this is definitely outside of my capabilities as I have zero experience with Capacitor.

**Comment by armenr** [comment]
One other question:

Is there any way to include or introduce a set of configuration entrypoints such that the CSS which quasar applies (when imported or included in a project) only applies to anything prefixed with `q-someComponent` as opposed to being applied universally?

**Reply by hawkeye64** [reply]
If you are using Vite, then you can import the quasar SASS in a wrapper, say `.my-quasar`, then any time you use a Quasar component, you will need to add to `class="my-quasar"` and formatting will only happen there.

**Comment by stefanvanherwijnen** [comment]
I finally managed to get SSG/SSR and some specific build features to work properly.
To prevent people from possibly assuming this is an official Quasar project, I renamed the it to @qso/app.

### Vite
All functionality is now handled by Vite and all Quasar support is provided by a plugin.
All configuration is provided through the `quasar` object in the Vite config.
This means that quasar.conf.js and quasar-variables.sass are now arguments to Vite itself through the `quasar` object.
The reason for this approach is that now you can create a Vite plugin which can alter or read anything in the build process.
Thus, app extensions can now be pure Vite plugins.
`extendQuasarConf` for example becomes:
```js
{
  name: 'app-extension-vite-plugin',
  config: (config, env) => {
    return {
      quasar: {
        conf: {
          // altered quasar.conf
        }
      }
    }
  }
}
```

### App extensions
For the Install/Uninstall API's I propose a composition approach. E.g. an app-extension package (e.g. my-app-extension) has an entry point ("main" in package.json) similar to:
```js
import { render } from 'quasar/app-extension'
export const InstallAPI = {
  render: ({ options }) => {
    render(...)
    // OR
    myOwnRenderFunction(...)
  }
}
```
The Install API then runs:
```js
const paths = ...
let { InstallAPI: { render } } = await import('my-app-extension')

render({
  paths,
  // anything else required for rendering
})
```
I.e. default functions are provided by Quasar and any specific options or configuration are supplied as an argument to the function.
This way developers have full freedom in their implementation, and the whole structure is more package manager friendly (i.e. no assumptions on where to load the InstallAPI (src/install.js)). In addition, the API's could have their own exports entry such that they can be imported as `my-app-extension/install` etc.

Any thoughts on this @hawkeye64?

I also changed the OP because most of it is outdated.

**Reply by hawkeye64** [reply]
I'm loving the direction this is going! ❤️ I'm on mobile right now, but as soon as I get on a computer I want to try this out. Keep up the good work.

**Comment by stefanvanherwijnen** [comment]
# Vitrify

It's been a while and in the mean time Vite support has arrived for @quasar/app. However, I required some very specific build functionalities so in the past months I decided to create a new build tool entirely based on the Vite ecosystem. I also ported some Quasar specific features such as SASS variables and app extensions. However, Quasar itself (the UI framework) is completely decoupled and all Quasar related functionality is contained within a Vitrify plugin (which is an extension of Vite plugins).

Vite's progress goes faster than I can keep up with, so Vitrify is a continuous work in progress but at this moment everything seems to work. The main goal is to have a single modular build tool which can be customized with plugins to provide any additional requirements.

## Features

- ESM only
- Fully compatible with Vite configuration.
- Integration with Fastify.
- Client-Side Rendering (CSR), Server-Side Rendering (SSR) and Fastify server build and development modes.
- Plugins for UI frameworks which can read their configuration from vitrify.config.js.
- A few helpful [configuration options](./packages/vitrify/src/node/vitrify-config.ts)
- Implements functional programming.
- Fully typed.

https://github.com/simsustech/vitrify

**Reply by IlCallo** [reply]
Cool! If you happen to find some patterns/improvements which could be useful for Quasar too, propose them and let's see if it makes sense to add them into Quasar too 👍🏻

**Reply by stefanvanherwijnen** [reply]
My main suggestion would be to fully switch to ESM and [kill CJS](https://github.com/vitejs/vite/issues/5839). 

For the CLI this is basically what Vitrify has grown into. For the UI this means breaking existing installations (https://github.com/quasarframework/quasar/issues/9157) 

Let me know if you'd like me to open an issue and create a proof of concept for Quasar (ui) as ES module.

**Reply by IlCallo** [reply]
As part of the team, I added a note into the Qv3 wish list to do a full switch to ESM
Personally, I don't think trying to kill CJS like this is a good thing, I think we'll need to support if for some time even after the whole codebase switches to ESM

**Comment by stefanvanherwijnen** [comment]
Killing CJS will make a lot of existing users unhappy so you are right. However, eventually CJS will be obsolete and in my opinion an ESM only project is much easier to maintain. 
Support is one of the things which makes Quasar great, but unfortunately it's hard to maintain older versions and the newest features at the same time. ("require is not defined" / "import syntax is not supported" 😫) 

I'd say, wait on Vite to kill CJS and follow them, because a large part of the ecosystem isn't ready yet at this moment.

**Reply by IlCallo** [reply]
Yes, I saw they'll probably use Vite 3 to transition to ESM, then stop supporting CJS entirely into Vite 4
